src/eunomia/server.py

```
@@ -18,24 +22,41 @@ class EunomiaServer:
     def __init__(self) -> None:
         self._engine = OpaPolicyEngine()
 
-    async def check_access(
-        self, principal_id: str, resource_id: str, access_method: str = "allow"
-    ) -> bool:
+    def _get_merged_attributes(self, entity: schemas.EntityAccess, db: Session) -> dict:
+        attributes = {}
+
+        if entity.uri is not None:
+            registered_attributes = crud.get_entity_attributes(entity.uri, db=db)
+            # if any attribute is colliding with the registered attributes, raise an error
+            for attribute in entity.attributes:
+                if (
+                    attribute.key in registered_attributes
+                    and registered_attributes[attribute.key] != attribute.value
+                ):
+                    raise ValueError(
+                        f"For entity '{entity.uri}', attribute '{attribute.key}' has a collision with the registered attributes"
+                    )
+            attributes.update(registered_attributes)
+
+        if entity.attributes:
+            attributes.update({item.key: item.value for item in entity.attributes})
```

Don't know if it might be better to return an error if the user send two new attributes that have the same keys. In this case only the last attribute value will be kept

---

Check similar comment in crud.py so we integrate the logic only in one place

---

Answering to the crud question also, I will add a validation for this in the schema